core/src/main/java/org/mule/runtime/core/internal/message/DefaultMessageBuilder.java

```diff
@@ -157,170 +155,170 @@ public DefaultMessageBuilder(org.mule.runtime.api.message.Message message) {
 
   @Override
   public CollectionBuilder itemMediaType(MediaType mediaType) {
-    if (dataType instanceof DefaultCollectionDataType) {
-      dataType =
-          ((DataTypeBuilder.DataTypeCollectionTypeBuilder) DataType.builder(this.dataType)).itemMediaType(mediaType).build();
+    if (payload.getDataType() instanceof DefaultCollectionDataType) {
+      payload = new TypedValue(payload.getValue(),
+                               ((DataTypeBuilder.DataTypeCollectionTypeBuilder) builder(payload.getDataType()))
+                                   .itemMediaType(mediaType).build());
     } else {
       throw new IllegalStateException("Item MediaType cannot be set, because payload is not a collection");
     }
     return this;
   }
 
   @Override
-  public InternalMessage.Builder attributes(TypedValue<?> typedValue) {
-    this.attributes = typedValue.getValue();
-    this.attributesDataType = typedValue.getDataType();
+  public InternalMessage.CollectionBuilder attributes(TypedValue<?> attributes) {
```

Comment:
@afelisatti had said there wasn't an issue with this as long as it wasn't so restrictive that it becomes unusable without IDE and she decided to leave these builders in place.  Here i'm fixing bug, rather than changing anything in the API.